README.md
carve
=====

A flexible 2.5 dimension distribution plot.  Let users carve the data!

# What is carve? 

Carve is a web-based visualization for diving into high dimensional data.  It is built on top of [d3](http://d3js.org/), [underscore](http://underscorejs.org/) and [sciencejs](https://github.com/jasondavies/science.js/). Applications can leverage [crossfilter](https://github.com/square/crossfilter) to quickly filter, sort, and group datasets based on events from carve, along with other interfaces.

Carve accepts numerical and categorical data.  Based on the types of data, it displays the distribution of data in one of three ways:

*Numerical vs. Numerical Scatterplot*

![Alt Scatterplot](http://rbkreisberg.github.com/carve/img/carve_scatterplot.png)

*Numerical vs Categorical KDE plot*

![Alt Scatterplot](http://rbkreisberg.github.com/carve/img/carve_kde.png)

*Categorical vs. Categorical Multiple Bar Chart*

![Alt Scatterplot](http://rbkreisberg.github.com/carve/img/carve_bar.png)

# What Else?

*Color* There is also the ability to color the data by a third dimension of the dataset.  The dimension must be a categorical feature (right now).  

*Highlight* The visualization can be told to highlight one of the values of the colored dimension (useful for external interfaces).

*Carving tools* Those panels above and to the left of the data plot allow the user to make precise cuts in one or two dimensions.  Cutting numerical data simply means locking a split in at a precision position by clicking on the panel.  Cutting categorical data is done by toggling values to create the desired split.

# Say What?

Check out a demonstration of carve: [Splitiscope](https://github.com/rbkreisberg/splitiscope)

# Use It!

Carve can be used as a global variable called 'carve' or as an AMD module using requirejs.  You can use multiple carves in the same page.

Carve depends on d3, underscore, and science.js.  Import these scripts prior to initializing a carve plot.

```javascript
var data = [

            { id: "Sparky", height : 10, weight : 14,  breed : "dachsund" },
            { id: "Lucy", height : 24, weight : 64,  breed : "retriever" },
            { id: "Chewbaca", height : 33, weight : 180, breed : "great dane" },
            { id: "Belle", height : 20, weight : 27,  breed : "whippet" },
            { id: "Crowe", height : 28, weight : 95,  breed : "ridgeback" },
            { id: "Chico", height : 7,  weight : 6,  breed : "chihuahua" },
            { id: "Katie", height : 32,  weight : 225,  breed : "mastiff" },
            { id: "Hamish", height : 27,  weight : 50,  breed : "saluki" }
]

var carve_obj = carve({
            width : 600,
            height: 450,
            radius : 20,
            margin : { top: 15, bottom: 20, left: 15, right: 50 }
        });
var carve_vis = carve_obj("#plot");

 carve_vis
        .colorBy({
            label : "breed",
            list : _.pluck(data, "breed")
        })
        .axisLabel({ x : "Weight (lbs)", y : "Height (in)" })
        .axisKey({ x : "weight", y : "height" })
        .data(data)
        .render();
```

There are additional settings that can be defined. The unique "id" key in each data point (defaults to "id"). The mapping between color_by class values and RGB Hex colors ( colorBy( { colors: colorList } ) ). Specific axis ranges (axisDomain), value dictionaries (axisValueDictionary), additional categorical values not in the data (axisInsistCategoricalValues), enforce axis data types (dataType).  Enable the "carving" behavior (enableCarving).

Additional settings using chained functions:
```

carve_vis
     .id("id")
     .colorBy({
            label : "breed",
            list : _.pluck(data, "breed"),
             colors : [ 
                    "#71C560",
                    "#9768C4",
                    "#98B8B8",
                    "#4F473D",
                    "#C1B14C",
                    "#B55381",
                    "#393939",
                    "#787878"
                ]
        })
        .axisLabel({ x : "Weight (lbs)", y : "Height (in)" })
        .axisKey({ x : "weight", y : "height" })
        .axisDomain({ x : [0,200], y : [0, 35 ] })
        .enableCarving(true)
        .data(data)
        .render();
```

Highlight data that have the "color_by" attributes of a specified value.  Remove the highlight.
```
carve_vis
    .highlight("saluki")
    .render();

carve_vis
    .highlight(null)
    .render();
```
<!-- 
    resources:
https://dl.dropboxusercontent.com/s/z0hyis11d003z80/carve.js
https://dl.dropboxusercontent.com/s/mjqt4svrh5mjbwv/carve.css
https://dl.dropboxusercontent.com/s/4xocll72aguws99/d3.js
https://dl.dropboxusercontent.com/s/hj6nj1d25ltax2m/science.v1.js -->

Check this out on JSFiddle: [demo](http://jsfiddle.net/rbkreisberg/bJnGL/)


CHANGELOG.md
2/7/2014
v0.1.0

Updates:
- width and height must be specified when carve is initialized as of now.
- simplied svg layout.  next version will have optional canvas element rendering for large datasets
- axis labels and split panels fit within padding + margin of plot
- y-axis tick labels are on the left side.  margin.left must accomodate the max length of these labels.
- split value labels now draw inside of the plot lightly.
- split panels/bars are thinner and hug the plot.

Todo:
- canvas rendering of large datasets
- 2d KDE plot of data when there is a lot of overdraw
- smarter tick formatting logic (no decimal points, fewest significant digits, etc)
- specify data point shape (circle, square, triangle..)
- autodetect parent element's height/width when possible.
- document API, events, css styling

v0.1.3
Updates:
- Bar charts render as a single rectangle

3/3/2014
v0.1.4
Updates:
- (carve).clear() wipes the plot away
- (carve).render() takes an optional callback function as a parameter.  the function is passed an <error> object that takes the form {error: true/false}

Todo
- canvas rendering of large datasets
- 2d KDE plot of data when there is a lot of overdraw
- smarter tick formatting logic (no decimal points, fewest significant digits, etc)
- specify data point shape (circle, square, triangle..)
- autodetect parent element's height/width when possible.
- document API, events, css styling


src/core.js
function carve(configObj) {

var config = configObj || {};

// useful defaults
var pointColors = [
                  "#71C560",
                  "#9768C4",
                  "#98B8B8",
                  "#4F473D",
                  "#C1B14C",
                  "#B55381"
            ];

//external accessible values
var __ = {
    width : 500,
    height : 400,
    margin : {
              "top" : 15,
              "bottom" : 15,
              "left" : 30,
              "right" : 10
            },
    radius : 4,
    dataType : {
                 "x" : 'n',
                 "y" : 'n',
                 "mix" : 'nn'
               },
    data : {x:[],y:[] },
    id : 'id',
    axisDomain : {
        "x" : "",
        "y" : ""
    },
    axisLabel : {
             "x" : "X Axis",
             "y" : "Y Axis"
           },
    axisValueDictionary : {
            "x" : {},
            "y" : {}
        },
    axisKey : {
            "x" : "x",
            "y" : "y"
        },
    axisInsistCategoricalValues : {
          "x" : [],
          "y" : []
        },
    highlight : null,
    colorFn : function(d) { return pointColors[0]; },
    colorBy : {
                label:"",
                list:[],
                colors: pointColors
              },
    clear : true,

    //Splitting related parameters
    enableCarving : false,
    splitColor : "#C1573B",
    splits : {},
    partition : {},
    partitionColors : [
                "#98B8B8",
                "#4F473D",
                "#C1B14C",
                "#B55381"
            ]
 };

  _.extend(__, config);

  var events = d3.dispatch.apply(this,["render", "resize", "highlight", "brush", "partitioncomplete"].concat(d3.keys(__))),
      outerWidth = function() { return __.width - __.margin.right - __.margin.left; },
      outerHeight = function() { return __.height - __.margin.top - __.margin.bottom; },
      plotWidth = function() { return outerWidth() - padding.right - padding.left; },
      plotHeight = function() { return outerHeight() - padding.top - padding.bottom; },
      displayWidth = function() { return plotWidth() - 20; },
      displayHeight = function() { return plotHeight() - 20; },

      scales = { x: d3.scale.ordinal(),
                y : {}
              },
      axisScales = { x: scales.x.copy(),
                  y: {}
                },
      kde = { x: null,
                y : null
              },

      selected = {x: null, y: null},
      min_uniq_points = 4,
      domainScalingFactor = 1.04,
      caseSplitsOpacityscale,
      padding = { top: 15, bottom: 15, left: 30, right: 20 },
      shapes = ['square','circle','cross','diamond','triangle-down','triangle-up'],
      symbolSize = Math.pow(__.radius,2),
      symbol = d3.svg.symbol().size(symbolSize).type(shapes[0]),
      symbolMap = d3.scale.ordinal().domain([0,5]).range(shapes),
      symbolFunction = _.compose(symbol.type, symbolMap),
      colorCategories = [],
      colorCategoryIsNumerical = false,
      strokeFunction = function(index) { return splitStrokeColors[index];},
      data_array = [],
      axisFn = {
          "x" :  d3.svg.axis().orient("bottom"),
          "y" : d3.svg.axis().orient("left")
        },
      update_duration = 300,
      bottom_surface, label_surface, split_surface, data_surface, partition_surface; // groups for axes, brushes

  var canvas = {},
         ctx = {};

      // Splitting-related variables
  var splitStrokeColors = ['red','green','black'],
      split_data = {x: {}, y: {}};

  // side effects for setters
  var side_effects = d3.dispatch.apply(this,d3.keys(__))
    .on("radius", function( ) { symbolSize = Math.pow(__.radius,2);})
    .on("data", function( ) { console.log('carve: data loaded');})
    .on("colorBy", setClassScales )
    .on("splits", parseSplits )
    .on("partition", setPartitions )
    .on("axisLabel", updateAxisLabels )
    .on("axisDomain", updateAxisDomains );
   
  var cv = function(selection) {
    selection = cv.selection = d3.select(selection);

    // __.width = 800; //selection[0][0].clientWidth;
    // __.height = 600; //selection[0][0].clientHeight;

    setAxes();

    //draw chart
    cv.svg = selection
                    .append('div')
                      .attr('class','cv_wrapper')
                    .append('svg')
                      .attr('class','cv')
                      .attr('width', __.width)
                      .attr('height', __.height);
    
    canvas['data'] = selection.select('div')
                    .append('canvas')
                    .attr('class','cv')
                    .style('position','absolute')
                    .attr('width',displayWidth())
                    .attr('height',displayHeight())
                    .style('left', (padding.left +__.margin.left) + 'px')
                    .style('top', (padding.top + __.margin.top) + 'px')
                    .style('z-index',-180)[0][0];
    
    ctx['data'] = canvas['data'].getContext("2d");

    var defs = cv.svg.append("defs");

                  defs.append("svg:clipPath")
                    .attr("id", "plot_clip")
                    .append("svg:rect")
                    .attr("id", "clip-rect")
                    .attr("x", "0")
                    .attr("y", "0")
                    .attr("width", displayWidth())
                    .attr("height", displayHeight());
            defs.append("svg:clipPath")
                    .attr("id", "viewbox_clip")
                    .append("svg:rect")
                    .attr("x", "0")
                    .attr("y", "0")
                    .attr("width", outerWidth())
                    .attr("height", outerHeight());

      label_surface = cv.svg.append('g')
                    .attr('class','label_surface')
                    .attr('transform','translate(' + (padding.left + __.margin.left + 10) + ',' +  (padding.top +  __.margin.top + 10 ) + ')');
    

    var plot_offset = cv.svg.append('svg')
                        .attr('clip','url(#viewbox_clip)')
                        .attr('overflow','hidden')
                        .attr('x',__.margin.left)
                        .attr('y', __.margin.top);
                       
    bottom_surface = plot_offset.append('g')
                        .attr('transform','translate(' + padding.left + ',' + padding.top + ')');

    partition_surface = bottom_surface.append('g')
                        .attr('transform','translate(10,10)')
                        .attr('class','partition_surface');
                         
    var top_surface = plot_offset.append('g')
                      .attr('transform','translate(' + padding.left + ',' + padding.top + ')');

    split_surface = top_surface.append('g')
                      .attr('transform','translate(10,0)')
                      .attr('class','split_surface');

    data_surface = top_surface.append('g')
                      .attr('transform','translate(10,10)')
                     .attr('clip-path','url(#plot_clip)');

    data_surface.append('g').attr('class','kde_surface');
    data_surface.append('g').attr('class','bar_surface');
    data_surface.append('g').attr('class','data');
    data_surface.append('g').attr('class','data_labels');

    //initialize colorBy class variables
    setClassScales();
    
    return cv;
  };

  cv.toString = function() { return "cv: (" + d3.keys(__.data[0]).length + " total) , " + __.data.length + " rows"; };

  // expose the state of the chart
  cv.state = __;
  
  // create getter/setters
  getset(cv, __, events);

  // expose events
  d3.rebind(cv, events, "on");

  // getter/setter with event firing
  function getset(obj,state,events)  {
    d3.keys(state).forEach(function(key) {
      obj[key] = function(x) {
        if (!arguments.length) return state[key];
        var old = state[key];
        state[key] = x;
        side_effects[key].call(cv,{"value": x, "previous": old});
        events[key].call(cv,{"value": x, "previous": old});
        return obj;
      };
    });
  }

cv.render = function(callback) {

  var err = parseData();
  if ( (err && err.error && err.error === true) || data_array.length <= 0) {
    if (_.isFunction(callback) ) {
      callback.call(this, err);
    }
    return this;
  }

  drawData();
  
  if (axisFn["y"].scale() !== undefined) drawAxes();
  if (_.isObject(split_data) && __.enableCarving ) {
    clearAllSplitSelections();
    clearAllSplitPointers();
    drawSplits();
    drawSplitLabel();
  }

  __.clear = false;

  if (_.isFunction(callback) ) {
      callback.call(this, err);
  }
  return this;
};

cv.resize = function() {

    cv.svg
            .attr('height',plotHeight())
            .attr('width',plotWidth())
            .attr("transform", "translate(" + __.margin.left + "," + __.margin.top + ")");
  return this;
};

function reRender() {
  updateAxes();
  drawData();
  if (__.enableCarving === false) { return; }
  drawSplits();
  drawPartitionSpans();
}

cv.clear = function() {
  clearDataPoints();
  clearBarPlots();
  clearDataLabels();
  clearKDE();
  clearAxes();
}